### PR TITLE
RavenDB-21064 - License Changes

### PR DESCRIPTION
**Pull request by grisha-kotler**
### Issue link

https://issues.hibernatingrhinos.com/issue/RavenDB-21064/License-limitations-for-v6.0

### Additional description

License changes starting from v6.0.

### Type of change

- New feature

### How risky is the change?

- Low 

### Backward compatibility

- Non breaking change

### Is it platform specific issue?

- No

### Documentation update

- This change requires a documentation update. Please mark the issue on YouTrack using `Documentation Required` tag.

### Testing by Contributor

- It has been verified by manual testing

### UI work

- It requires further work in the Studio. Please mark the issue on YouTrack using `Studio Required` tag.

### CODE REVIEW

File: src/Raven.Server/Documents/Handlers/Processors/Indexes/AbstractIndexHandlerProcessorForSuggestIndexMerge.cs [hunk 1]
```diff
@@ -14,6 +15,10 @@ internal abstract class AbstractIndexHandlerProcessorForSuggestIndexMerge<TReque
 {
     protected AbstractIndexHandlerProcessorForSuggestIndexMerge([NotNull] TRequestHandler requestHandler) : base(requestHandler)
     {
+        if (requestHandler.ServerStore.LicenseManager.LicenseStatus.HasIndexCleanup)
```
Comment (kalczur) [inline src/Raven.Server/Documents/Handlers/Processors/Indexes/AbstractIndexHandlerProcessorForSuggestIndexMerge.cs]:
```suggestion
        if (!requestHandler.ServerStore.LicenseManager.LicenseStatus.HasIndexCleanup)
```

**Review by kalczur — CHANGES_REQUESTED** [review]

File: src/Raven.Server/Documents/Handlers/Processors/Indexes/AbstractIndexHandlerProcessorForSuggestIndexMerge.cs (hunk 1 above)
Comment (grisha-kotler) [reply src/Raven.Server/Documents/Handlers/Processors/Indexes/AbstractIndexHandlerProcessorForSuggestIndexMerge.cs]: correct. thanks @kalczur

File: src/Raven.Server/Commercial/LicenseManager.cs [hunk 2]
```diff
@@ -433,6 +433,36 @@ public async Task ActivateAsync(License license, string raftRequestId, bool skip
...
+                {
+                    var updatedLicense = await GetUpdatedLicenseForActivation(license);
+                    if (updatedLicense != null)
+                    {
+                        await ActivateAsync(updatedLicense, raftRequestId, skipGettingUpdatedLicense: true);
+                        return;
+                    }
+                }
+
+                //TODO: if (updatedLicense == null)
```
Comment (ppekrol) [inline src/Raven.Server/Commercial/LicenseManager.cs]: TODO, remember about this

File: src/Raven.Server/Commercial/LicenseManager.cs [hunk 3]
```diff
@@ -433,6 +433,36 @@ public async Task ActivateAsync(License license, string raftRequestId, bool skip
...
+                        $"License already expired on: {licenseStatus.FormattedExpiration} and we failed to get an updated one from {ApiHttpClient.ApiRavenDbNet}.";
+                    if (licenseStatus.IsIsv)
+                    {
+                        errorMessage += $" Since this is an ISV license, you can use this license with any RavenDB version that was released prior to {licenseStatus.FormattedExpiration}";
+                    }
+
+                    throw new LicenseLimitException(errorMessage);
+                }
+
+                throw new LicenseLimitException($"Your license version is {licenseStatus.Version}. " +
```
Comment (ppekrol) [inline src/Raven.Server/Commercial/LicenseManager.cs]: We need a better message here with a link to the tool

File: src/Raven.Server/Program.cs [hunk 4]
```diff
@@ -319,82 +320,122 @@ public static unsafe int Main(string[] args)
...
+                    serverStore.LicenseManager.OnBeforeInitialize += () =>
+                        serverStore.LicenseManager.TryActivateLicenseAsync(throwOnActivationFailure: serverStore.Server.ThrowOnLicenseActivationFailure)
+                            .Wait(serverStore.ServerShutdown);
+                    return;
                 }
+            }
 
-                var errorMessage = $"Cannot start the RavenDB server because the expiration date of current license ({FormattedDateTime(licenseStatus.Expiration ?? DateTime.MinValue)}) " +
-                                   $"is before the release date of this version ({FormattedDateTime(RavenVersionAttribute.Instance.ReleaseDate)})";
+            //TODO: add link and explain what need to be done
```
Comment (ppekrol) [inline src/Raven.Server/Program.cs]: TODO

File: src/Raven.Server/ServerWide/ClusterStateMachine.cs [hunk 5]
```diff
@@ -581,6 +614,7 @@ protected override void Apply(ClusterOperationContext context, BlittableJsonRead
                     case nameof(PutServerWideBackupConfigurationCommand):
                         var serverWideBackupConfiguration = UpdateValue<ServerWideBackupConfiguration>(context, type, cmd, index, skipNotifyValueChanged: true);
                         UpdateDatabasesWithServerWideBackupConfiguration(context, type, serverWideBackupConfiguration, index);
+                        AssertServerWideTasks(serverStore);
```
Comment (ppekrol) [inline src/Raven.Server/ServerWide/ClusterStateMachine.cs]: we are first putting the configuration then asserting?

File: src/Raven.Server/ServerWide/ClusterStateMachine.cs [hunk 6]
```diff
@@ -746,6 +783,12 @@ void AddUnrecoverableNotification(string error, Exception exception)
             }
         }
 
+        private static void AssertServerWideTasks(ServerStore serverStore)
```
Comment (ppekrol) [inline src/Raven.Server/ServerWide/ClusterStateMachine.cs]: move all this licensing code to a partial

File: src/Raven.Server/ServerWide/ClusterStateMachine.cs [hunk 7]
```diff
@@ -2698,6 +2754,393 @@ private void UpdateDatabase(ClusterOperationContext context, string type, Blitta
             }
         }
 
+        private void AssertSubscriptionsLicenseLimits(ServerStore serverStore, Table items, PutSubscriptionCommand command, ClusterOperationContext context)
```
Comment (ppekrol) [inline src/Raven.Server/ServerWide/ClusterStateMachine.cs]: move to partial

File: src/Raven.Server/ServerWide/ClusterStateMachine.cs [hunk 8]
```diff
@@ -2698,6 +2754,393 @@ private void UpdateDatabase(ClusterOperationContext context, string type, Blitta
...
+        {
+            var maxSubscriptionsPerDatabase = serverStore.LicenseManager.LicenseStatus.MaxNumberOfSubscriptionsPerDatabase;
+            if (maxSubscriptionsPerDatabase != null && maxSubscriptionsPerDatabase >= 0 && maxSubscriptionsPerDatabase < GetSubscriptionsCountForDatabase(command.DatabaseName))
+            {
+                if (CanAssertLicenseLimits(context, minBuildVersion: 60_000) == false)
```
Comment (ppekrol) [inline src/Raven.Server/ServerWide/ClusterStateMachine.cs]: in that partial put a cost with this number

File: src/Raven.Server/ServerWide/ClusterStateMachine.cs [hunk 9]
```diff
@@ -2698,6 +2754,393 @@ private void UpdateDatabase(ClusterOperationContext context, string type, Blitta
...
+                    return false;
+
+                if (limit.Value.BuildInfo.BuildVersion < minBuildVersion)
+                    return false;
+            }
+
+            return true;
+        }
+
+        private enum TotalType
```
Comment (ppekrol) [inline src/Raven.Server/ServerWide/ClusterStateMachine.cs]: we need a better name, total type of what?

File: src/Raven.Server/ServerWide/ClusterStateMachine.cs [hunk 10]
```diff
@@ -2744,6 +3186,7 @@ private static void UpdateIndexForBackup(DatabaseRecord databaseRecord, string t
                 case nameof(EditTimeSeriesConfigurationCommand):
                 case nameof(PutAutoIndexCommand):
                 case nameof(PutDatabaseClientConfigurationCommand):
+                case nameof(EditDatabaseClientConfigurationCommand):
```
Comment (ppekrol) [inline src/Raven.Server/ServerWide/ClusterStateMachine.cs]: this was ordered alphabetically for easier sync :)

File: src/Raven.Server/Commercial/LicenseManager.cs [hunk 11]
```diff
@@ -433,6 +433,36 @@ public async Task ActivateAsync(License license, string raftRequestId, bool skip
...
+                    }
+                }
+
+                //TODO: if (updatedLicense == null)
+                {
+                    var errorMessage =
+                        $"License already expired on: {licenseStatus.FormattedExpiration} and we failed to get an updated one from {ApiHttpClient.ApiRavenDbNet}.";
+                    if (licenseStatus.IsIsv)
+                    {
+                        errorMessage += $" Since this is an ISV license, you can use this license with any RavenDB version that was released prior to {licenseStatus.FormattedExpiration}";
```
Comment (ppekrol) [inline src/Raven.Server/Commercial/LicenseManager.cs]: is word 'any' valid?

File: src/Raven.Server/ServerWide/ClusterStateMachine.cs (hunk 6 above)
Comment (grisha-kotler) [reply src/Raven.Server/ServerWide/ClusterStateMachine.cs]: done

File: src/Raven.Server/ServerWide/ClusterStateMachine.cs (hunk 7 above)
Comment (grisha-kotler) [reply src/Raven.Server/ServerWide/ClusterStateMachine.cs]: done

File: src/Raven.Server/ServerWide/ClusterStateMachine.cs (hunk 8 above)
Comment (grisha-kotler) [reply src/Raven.Server/ServerWide/ClusterStateMachine.cs]: done

File: src/Raven.Server/ServerWide/ClusterStateMachine.cs (hunk 9 above)
Comment (grisha-kotler) [reply src/Raven.Server/ServerWide/ClusterStateMachine.cs]: done

File: src/Raven.Server/ServerWide/ClusterStateMachine.cs (hunk 10 above)
Comment (grisha-kotler) [reply src/Raven.Server/ServerWide/ClusterStateMachine.cs]: it's easier to search commands when they are near each other :)
reverted.

File: src/Raven.Server/ServerWide/ClusterStateMachine.cs (hunk 5 above)
Comment (grisha-kotler) [reply src/Raven.Server/ServerWide/ClusterStateMachine.cs]: done

File: src/Raven.Server/ServerWide/ClusterStateMachine.cs (hunk 10 above)
Comment (ppekrol) [reply src/Raven.Server/ServerWide/ClusterStateMachine.cs]: Unless after some time half of the commands that should be here is missing and you need to figure out which ones are those :)

**Comment by kalczur** [comment]
I'm working on fitting new fields into the studio

File: src/Raven.Server/Commercial/LicenseManager.cs [hunk 12]
```diff
@@ -148,7 +148,7 @@ public void Initialize(StorageEnvironment environment, TransactionContextPool co
             {
                 _leaseLicenseTimer = new Timer(state =>
                     AsyncHelpers.RunSync(ExecuteTasks), null,
-                    (int)TimeSpan.FromMinutes(1).TotalMilliseconds,
+                    (int)TimeSpan.FromMinutes(100).TotalMilliseconds, //TODO: revert
```
Comment (ppekrol) [inline src/Raven.Server/Commercial/LicenseManager.cs]: TODO

**Review by ppekrol — APPROVED** [review]

**Review by ml054 — APPROVED** [review]

File: src/Raven.Server/Commercial/LicenseManager.cs (hunk 2 above)
Comment (grisha-kotler) [reply src/Raven.Server/Commercial/LicenseManager.cs]: done

File: src/Raven.Server/Commercial/LicenseManager.cs (hunk 11 above)
Comment (grisha-kotler) [reply src/Raven.Server/Commercial/LicenseManager.cs]: removed

File: src/Raven.Server/Commercial/LicenseManager.cs (hunk 3 above)
Comment (grisha-kotler) [reply src/Raven.Server/Commercial/LicenseManager.cs]: done

File: src/Raven.Server/Commercial/LicenseStatus.cs [hunk 13]
```diff
@@ -259,7 +325,22 @@ public DynamicJsonValue ToJson()
...
+                [nameof(HasPeriodicBackup)] = HasPeriodicBackup,
+                [nameof(HasClientConfiguration)] = HasClientConfiguration,
+                [nameof(HasStudioConfiguration)] = HasStudioConfiguration,
                 [nameof(HasQueueSink)] = HasQueueSink,
+                [nameof(HasDataArchival)] = HasDataArchival,
+                [nameof(HasRevisionsInSubscriptions)] = HasRevisionsInSubscriptions,
+                [nameof(HasMultiNodeSharding)] = HasMultiNodeSharding,
+                [nameof(MaxNumberOfRevisionsToKeep)] = MaxNumberOfRevisionsToKeep,
+                [nameof(MaxNumberOfRevisionAgeToKeepInDays)] = MaxNumberOfRevisionAgeToKeepInDays,
+                [nameof(MinPeriodForExpirationInHours)] = MaxNumberOfRevisionAgeToKeepInDays,
```
Comment (kalczur) [inline src/Raven.Server/Commercial/LicenseStatus.cs]:
```suggestion
                [nameof(MinPeriodForExpirationInHours)] = MinPeriodForExpirationInHours,
```

**Review by kalczur — CHANGES_REQUESTED** [review]

File: src/Raven.Server/Commercial/LicenseStatus.cs (hunk 13 above)
Comment (grisha-kotler) [reply src/Raven.Server/Commercial/LicenseStatus.cs]: done

File: src/Raven.Server/Program.cs (hunk 4 above)
Comment (grisha-kotler) [reply src/Raven.Server/Program.cs]: done

File: src/Raven.Server/ServerWide/ClusterStateMachine.cs [hunk 14]
```diff
@@ -1775,6 +1797,14 @@ private unsafe List<string> AddDatabase(ClusterOperationContext context, Blittab
                     using (var databaseRecordAsJson = UpdateDatabaseRecordIfNeeded(databaseExists, shouldSetClientConfigEtag, index, addDatabaseCommand, newDatabaseRecord, context))
                     {
                         UpdateValue(index, items, valueNameLowered, valueName, databaseRecordAsJson);
+
+                        AssertLicenseLimits(type, serverStore, addDatabaseCommand.Record, items, context);
+
+                        foreach (var command in _licenseLimitsCommandsForCreateDatabase)
+                        {
+                            AssertLicenseLimits(command, serverStore, addDatabaseCommand.Record, items, context);
+                        }
```
Comment (karmeli87) [inline src/Raven.Server/ServerWide/ClusterStateMachine.cs:1806]: `LicenseLimitException` is an _expected_ exception, so the tx will be committed... 
In the line above you already do `UpdateValue` and put the database, so I expect it to throw and _accept_ the command.

But maybe I'm wrong :-)

File: src/Raven.Server/ServerWide/ClusterStateMachine.cs [hunk 15]
```diff
@@ -2684,6 +2714,8 @@ private void UpdateDatabase(ClusterOperationContext context, string type, Blitta
                     UpdateIndexForBackup(databaseRecord, type, index);
                     var updatedDatabaseBlittable = DocumentConventions.DefaultForServer.Serialization.DefaultConverter.ToBlittable(databaseRecord, context);
                     UpdateValue(index, items, valueNameLowered, valueName, updatedDatabaseBlittable);
+
+                    AssertLicenseLimits(type, serverStore, databaseRecord, items, context);
```
Comment (karmeli87) [inline src/Raven.Server/ServerWide/ClusterStateMachine.cs:2718]: same issue here I guess

**Pull request merged**